### PR TITLE
add nodouble option for all cwise ops

### PR DESCRIPTION
**Pull request by guoyejun**
this is the following patch of ac98d1184008e4 to support nodouble
option for all cwise ops. The macros REGISTER* defined within
__ANDROID_TYPES_SLIM__ should be changed to empty, and it impacts
all the cwise ops, so, all the changes for the cwise ops have to be
in a single patch.

there will be more patches for other ops to support nodouble option.

### CODE REVIEW

File: tensorflow/core/kernels/cwise_ops_common.h [hunk 1]
```diff
@@ -476,17 +476,17 @@ struct ApproximateEqual<CPUDevice, T> {
...
-#define REGISTER4(OP, D, N, F, T0, T1, T2, T3) REGISTER(OP, D, N, F, T0)
-#define REGISTER5(OP, D, N, F, T0, T1, T2, T3, T4) REGISTER(OP, D, N, F, T0)
-#define REGISTER6(OP, D, N, F, T0, T1, T2, T3, T4, T5) REGISTER(OP, D, N, F, T0)
-#define REGISTER7(OP, D, N, F, T0, T1, T2, T3, T4, T5, T6) \
-  REGISTER(OP, D, N, F, T0)
-#define REGISTER8(OP, D, N, F, T0, T1, T2, T3, T4, T5, T6, T7) \
-  REGISTER(OP, D, N, F, T0)
-#define REGISTER9(OP, D, N, F, T0, T1, T2, T3, T4, T5, T6, T7, T8) \
-  REGISTER(OP, D, N, F, T0)
+// the float data type is registered with TF_CALL_SYCL_NUMBER_TYPES,
```
Comment (vrv) [inline tensorflow/core/kernels/cwise_ops_common.h]: I don't really like that we're relying on a SYCL define for ANDROID_TYPES_SLIM, which are two different systems.  Can we make it so that SYCL and ANDROID_TYPES_SLIM don't interfere in any way?

Comment (guoyejun) [inline tensorflow/core/kernels/cwise_ops_common.h]: thanks for pointing it, i realized that i had a misunderstanding that REGISTER* is only for sycl backend, actually, it is for all the backends cpu, cuda and sycl.  My change relative to ANDROID_TYPES_SLIM is not right. I'll send new patches.

**Comment by guoyejun** [comment]
thanks @vrv pointing my mistake, i'll rewrite the code to add new macros such as REGISTER_SYCL, and REGISTER6_SYCL. In the new code, the change could be separated to be one patch for one op, it could make each patch simple and easy to be reviewed.  So, my plan is to close this PR, and send out new PRs one by one (send out the second patch only after the first one is accepted).

**Comment by vrv** [comment]
I'd be fine with all the changes in the same PR actually, since they are logically all the same, and can be tested all at once (less load on our test infrastructure). They are mostly mechanical changes so it's easier to review in a single batch :)

**Comment by guoyejun** [comment]
got it, i'll continue this PR with new patches.

**Comment by rmlarsen** [comment]
@vrv Could you take another look, please?

**Comment by vrv** [comment]
I think @benoitsteiner probably wants to look at this since it's SYCL related.

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by guoyejun** [comment]
hello, i checked the failure log but do not understand how it relative to my change, could anyone help? thanks.

the last few lines of the log:
PASS: //tensorflow/python:session_benchmark
PASS: //tensorflow/python:matmul_benchmark_test
Build was aborted
Aborted by unknown
[Set GitHub commit status (universal)] PENDING on repos [] (sha:ef4e31e) with context:tensorflow-pull-requests-gpu
Unable to get pull request builder trigger!!
Setting status of 12276a750d6240f93878f84fa125f6be48566e7d to FAILURE with url https://ci.tensorflow.org/job/tensorflow-pull-requests-gpu/6269/ and message: 'FAILURE
 '
Using context: Linux GPU
Finished: ABORTED

**Comment by rmlarsen** [comment]
@guoyejun The failure was unrelated to your change.

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

File: tensorflow/core/kernels/cwise_op_abs.cc [hunk 2]
```diff
@@ -38,7 +38,11 @@ REGISTER_KERNEL_BUILDER(Name("Abs")
 #endif
 
 #if TENSORFLOW_USE_SYCL
-REGISTER3(UnaryOp, SYCL, "Abs", functor::abs, float, double, int64);
+#define REGISTER_KERNEL(type) REGISTER(UnaryOp, SYCL, "Abs", functor::abs, type);
```
Comment (benoitsteiner) [inline tensorflow/core/kernels/cwise_op_abs.cc]: Can you create a new macro that call # define REGISTER_KERNEL(type) TF_CALL_SYCL_NUMBER_TYPES(REGISTER_KERNEL) in a central location and also make sure that the new macro undefines REGISTER_KERNEL when it's done. That will avoid a lot of boilerplate code.

File: tensorflow/core/kernels/cwise_ops_sycl_common.h [hunk 3]
```diff
@@ -154,6 +154,28 @@ struct BinaryFunctor<SYCLDevice, Functor, NDIMS, has_errors> {
   DEFINE_BINARY5(F, T5, T6, T7, T8, T9)
 
 }  // end namespace functor
+
+
+#if defined(__ANDROID_TYPES_SLIM__)
```
Comment (benoitsteiner) [inline tensorflow/core/kernels/cwise_ops_sycl_common.h]: Why is this needed for coefficient wise operations and not for all the other sycl kernels ?

**Review by benoitsteiner — CHANGES_REQUESTED** [review]
I'm not sure this is a scalable approach.

File: tensorflow/core/kernels/cwise_op_abs.cc (hunk 2 above)
Comment (guoyejun) [inline tensorflow/core/kernels/cwise_op_abs.cc]: yes, good idea.

File: tensorflow/core/kernels/cwise_ops_sycl_common.h (hunk 3 above)
Comment (guoyejun) [inline tensorflow/core/kernels/cwise_ops_sycl_common.h]: In the current code, it is needed for cwise ops, see the code at the end of cwise_ops_common.h,  this change is to keep this logic.

As for other sycl kernels, i'll review it when I face them. As for this patch, it just focuses on cwise ops.

File: tensorflow/core/kernels/cwise_op_abs.cc (hunk 2 above)
Comment (guoyejun) [inline tensorflow/core/kernels/cwise_op_abs.cc]: i tried it, but did not found a way to implement it. My code looks like below, while it is illegal grammar.  Do you have any idea @benoitsteiner , thanks.

#define REGISTER_NUMBER_SYCL(OP, N, F)          \
  #define REGISTER_SYCL_KERNEL(type) REGISTER(OP, SYCL, N, F, type)   \
  TF_CALL_SYCL_NUMBER_TYPES(REGISTER_SYCL_KERNEL);                    \
  #undef REGISTER_SYCL_KERNEL

**Comment by martinwicke** [comment]
@benoitsteiner Can you take a look at the replies?

**Comment by drpngx** [comment]
@benoitsteiner what do you think?

**Comment by sb2nov** [comment]
@benoitsteiner ping?

**Comment by drpngx** [comment]
@benoitsteiner any luck with this?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @benoitsteiner: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @benoitsteiner: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by guoyejun** [comment]
just close it since no feedback for a long time, and code base change results into conflicts
